Modelica/Electrical/Analog/Ideal/ImprovedOpAmpLimited.mo

```diff
@@ -0,0 +1,110 @@
+model ImprovedOpAmpLimited "Improved operational amplifier with limitation"
+  parameter Real V0=15000.0 "No-load amplification";
+  parameter SI.Time Tr=1e-5 "Rise time 10%-90% of output voltage";
+  parameter Boolean useSupply=false
+    "Use supply pins (otherwise constant supply)" annotation (Evaluate=true);
+  parameter SI.Voltage Vps=+15 "Positive supply voltage"
+    annotation (Dialog(enable=not useSupply));
+  parameter SI.Voltage Vns=-15 "Negative supply voltage"
+    annotation (Dialog(enable=not useSupply));
+  parameter Modelica.Electrical.Analog.Types.InitOpAmp initOpAmp=Modelica.Electrical.Analog.Types.InitOpAmp.Linear
+    "Initialization of rise time fristOrder"
+    annotation (Evaluate=true, Dialog(group="Initialization"));
+  Boolean satPos=v_int>=vps "Indicates positive Saturation";
+  Boolean satNeg=v_int<=vns "Indicates negative Saturation";
+  SI.Voltage vps "Positive supply voltage";
+  SI.Voltage vns "Negative supply voltage";
+  SI.Voltage v_in(start=0)=in_p.v - in_n.v "Input voltage difference";
+  SI.Voltage v_int "Intermediate voltage";
+  SI.Voltage v_out=out.v "Output voltage to ground";
+  SI.Current i_out(start=0)=out.i "Output current into OpAmp";
+  SI.Power p_in=in_p.v*in_p.i + in_n.v*in_n.i "Input power";
+  SI.Power p_out=out.v*out.i "Output power";
+  SI.Power p_s=-(p_in + p_out) "Supply power";
+  SI.Current i_s=p_s/(vps - vns) "Supply current";
+  Modelica.Electrical.Analog.Interfaces.PositivePin in_p
+    "Positive pin of the input port" annotation (Placement(transformation(
+          extent={{-90,-70},{-110,-50}})));
+  Modelica.Electrical.Analog.Interfaces.NegativePin in_n
+    "Negative pin of the input port" annotation (Placement(transformation(
+          extent={{-110,50},{-90,70}})));
+  Modelica.Electrical.Analog.Interfaces.PositivePin out
+    "Pin of the output port" annotation (Placement(transformation(extent={{
+            110,-10},{90,10}}), iconTransformation(extent={{110,-10},
+            {90,10}})));
+  //optional supply pins
+  Modelica.Electrical.Analog.Interfaces.PositivePin s_p(final i=+i_s, final v=vps)
+    if useSupply "Optional positive supply pin" annotation (Placement(
+        transformation(extent={{10,90},{-10,110}})));
+  Modelica.Electrical.Analog.Interfaces.NegativePin s_n(final i=-i_s, final v=vns)
+    if useSupply "Optional negative supply pin" annotation (Placement(
+        transformation(extent={{-10,-110},{10,-90}})));
+initial equation
+  if initOpAmp == Modelica.Electrical.Analog.Types.InitOpAmp.UpperLimit then
+    v_int=vps;
+  elseif initOpAmp == Modelica.Electrical.Analog.Types.InitOpAmp.LowerLimit
+       then
+    v_int=vns;
+  else
+    v_int=V0*v_in;
+  end if;
+equation
+  if not useSupply then
+    vps = Vps;
+    vns = Vns;
+  end if;
+  assert(vps > vns, "Supply voltages are not consistent: vps <= vns");
+  //input currents
+  in_p.i = 0;
+  in_n.i = 0;
+  //firstOrder to model Rise time
+  der(v_int) = (V0*v_in - v_int)/(Tr/log(9));
```

Comment:
I don't fully get the log(9), and I think it should be explained/replaced by some final parameter.
I assume it is related to 10-90% rise time.

I also wonder if it would be possible to have an option to de-activate the firstOrder part.

Reply:
Yes we could use a choice between fristOrder and simple gain - I'll try that,
Maybe dependent on the size of the time constant (too large is bad, too small is bad).
And I'll name the time constant differently and remove the log(9), this is not exact.
The log(9) stems from:
0.1 = (1 - exp(-t10/Tau)
0.9 = (1 - exp(-t90/Tau)
Tr = t90 - t10 = Tau*ln(9)

Reply:
@HansOlsson I think it is a good idea to have an option to deavtivate the firstOrder part. 
On the other hand this makes initialization much more complicated. 
With firstOrder it's easy: Just initialize the output of the firstOrder.
With simple gain I'd have to fiddle around with homotopy (to be able to fix the output at positive or negative limit) which isn't that easy when limits are not constant (parameters).